backend-shared/config/db/util/application_info_cache.go

```diff
@@ -0,0 +1,411 @@
+		}
+	} else {
+		// An error occurred, so remove the cache entry from both the application state, and the application,
+		// so that it can be re-acquired.
+		// A common error returned by dbQueries above, will be that the applicate state cannot be updated, because the application no
+		// longer exists in the database.
+		// This is normal, and occurs when a GitOpsDeployment is deleted (and then the corresponding Application/Application state are deleted as well).
+		// In this case, we need to remove the Application/ApplicationState from the DB, as they have likely been deleted (and thus should no longer be cached.)
+		delete(cacheApp, req.createOrUpdateAppStateObject.Applicationstate_application_id)
+		delete(cacheAppState, req.createOrUpdateAppStateObject.Applicationstate_application_id)
+	}
+
+	req.responseChannel <- applicationInfoCacheResponse{
+		err: err,
+	}
+
+}
+
+func processUpdateAppStateMessage(dbQueries db.DatabaseQueries, req applicationInfoCacheRequest, cacheApp map[string]applicationCacheEntry, cacheAppState map[string]applicationStateCacheEntry) {
+
+	err := dbQueries.UpdateApplicationState(req.ctx, &req.createOrUpdateAppStateObject)
+
+	if err == nil {
+		// Update the cache on success
+		var appState db.ApplicationState = req.createOrUpdateAppStateObject
+		if entryInCache, exists := cacheAppState[req.primaryKey]; !exists {
+			entryInCache.appState = appState
+			cacheAppState[req.primaryKey] = entryInCache
+		}
+	} else {
+		delete(cacheApp, req.createOrUpdateAppStateObject.Applicationstate_application_id)
+		delete(cacheAppState, req.createOrUpdateAppStateObject.Applicationstate_application_id)
+	}
+
+	req.responseChannel <- applicationInfoCacheResponse{
+		err: err,
+	}
+
+}
+
+func processDeleteAppStateMessage(dbQueries db.DatabaseQueries, req applicationInfoCacheRequest, cacheApp map[string]applicationCacheEntry, cacheAppState map[string]applicationStateCacheEntry) {
+
+	if db.IsEmpty(req.primaryKey) {
+		log.Error(fmt.Errorf("PrimaryKey should not be nil"), "SEVERE: "+req.primaryKey+" not found")
+	}
+
+	// Remove from cache
+	delete(cacheAppState, req.primaryKey)
+	delete(cacheApp, req.primaryKey)
+
+	// Remove from DB
+	rowsAffected, err := dbQueries.DeleteApplicationStateById(req.ctx, req.primaryKey)
+
+	req.responseChannel <- applicationInfoCacheResponse{
+		rowsAffectedForDelete: rowsAffected,
+		err:                   err,
+	}
+
+}
+
+func processGetAppStateMessage(dbQueries db.DatabaseQueries, req applicationInfoCacheRequest, cacheApp map[string]applicationCacheEntry, cacheAppState map[string]applicationStateCacheEntry) {
+
+	appState := db.ApplicationState{
+		Applicationstate_application_id: req.primaryKey,
+	}
+
+	var err error
+	var valueFromCache bool
+
+	if db.IsEmpty(req.primaryKey) {
+		log.Error("PrimaryKey should not be nil: " + req.primaryKey + " not found")
+	} else {
+		delete(cacheAppState, req.createOrUpdateAppStateObject.Applicationstate_application_id)
```

Comment:
These two lines in the else branch of this if statement will cause the cache to be cleared after each GetAppStateMessage (which defeats the purpose of a cache).

Presuming I understand/recall the logic correctly, I think these deletes should move below.